src/tools/compiletest/src/lib.rs

```diff
@@ -545,46 +543,62 @@ pub fn test_opts(config: &Config) -> test::TestOpts {
     }
 }
 
+/// Read-only context data used during test collection.
+struct TestCollectorCx {
+    config: Arc<Config>,
+    cache: HeadersCache,
+    common_inputs_stamp: Stamp,
+    modified_tests: Vec<PathBuf>,
+}
+
+/// Mutable state used during test collection.
+struct TestCollector {
+    tests: Vec<test::TestDescAndFn>,
+    found_path_stems: HashSet<PathBuf>,
```

Comment:
Suggestion: could this be a `BTreeSet`? Perf on this specifically isn't very important, but I do like a stable order if we ever need to debug print this and such.

Reply:
It used to be a BTreeSet, and was converted to a HashSet for perf: #120273.

(I don't personally know whether that was justified, but either way I'd prefer to leave it as-is for this PR.)

Note that the code that prints an error message does sort the list of collisions before printing them.

Reply:
That linked PR said "wall time didnt improve", let's just use BTreeSet

Reply:
You know what, during the collection phase this could just be a list, since exact duplicates should never happen anyway.

Then the check function can convert it to whatever set representation it wants in one go.

Reply:
That's fine with me too.

Reply:
On the other hand, a vector is still subject to the arbitrary order of directory traversal. So I'll just go back to BTreeMap, since there's no observable perf difference.